File: code/day_021/02_将有序数组转换为二叉搜索树/v3_自己正确.cpp
class Solution {
public:
    TreeNode* ArrayToBST(vector<int>& nums, int left, int right) {
        if(left > right) {
            return NULL;
        }

        int mid = left + (right-left) / 2;

        TreeNode* node = new TreeNode(nums[mid]);
        node->left = ArrayToBST(nums, left, mid-1);
        node->right = ArrayToBST(nums, mid+1, right);

        return node;
    }

    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if(nums.size() == 0) return NULL;
        
        TreeNode* node = new TreeNode(nums.size() / 2);

        node = ArrayToBST(nums, 0, nums.size()-1);

        return node;
    }
};

File: leetcode/code/day_002_双指针/03_三数之和/v3_自己正确.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        result = []
        nums.sort()

        for i in range(len(nums) - 2):
            # 第一个数去重
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left = i + 1
            right = len(nums) - 1

            while left < right:
                s = nums[i] + nums[left] + nums[right]

                if s == 0:
                    result.append([nums[i], nums[left], nums[right]])

                    # 第二个数去重
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1

                    # 第三个数去重
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1

                    left += 1
                    right -= 1

                elif s < 0:
                    left += 1
                else:
                    right -= 1

        return result